core/node/da_dispatcher/src/da_dispatcher.rs

```diff
@@ -27,178 +36,296 @@ impl DataAvailabilityDispatcher {
         config: DADispatcherConfig,
         client: Box<dyn DataAvailabilityClient>,
     ) -> Self {
+        let request_semaphore = Arc::new(tokio::sync::Semaphore::new(100));
         Self {
             pool,
             config,
             client,
+            request_semaphore,
         }
     }
 
-    pub async fn run(self, mut stop_receiver: Receiver<bool>) -> anyhow::Result<()> {
-        loop {
-            if *stop_receiver.borrow() {
-                break;
-            }
+    pub async fn run(self, stop_receiver: Receiver<bool>) -> anyhow::Result<()> {
+        let subtasks = futures::future::join(
+            async {
+                if let Err(err) = self.dispatch_batches(stop_receiver.clone()).await {
+                    tracing::error!("dispatch error {err:?}");
+                }
+            },
+            async {
+                if let Err(err) = self.inclusion_poller(stop_receiver.clone()).await {
+                    tracing::error!("poll_for_inclusion error {err:?}");
+                }
+            },
+        );
+
+        tokio::select! {
+            _ = subtasks => {},
+        }
+        Ok(())
+    }
+
+    async fn dispatch_batches(&self, stop_receiver: Receiver<bool>) -> anyhow::Result<()> {
+        let (tx, mut rx) = mpsc::channel(100);
+
+        let next_expected_batch = Arc::new(AtomicI64::new(NO_FIRST_BATCH));
+
+        let stop_receiver_clone = stop_receiver.clone();
+        let pool_clone = self.pool.clone();
+        let config_clone = self.config.clone();
+        let next_expected_batch_clone = next_expected_batch.clone();
+        let pending_blobs_reader = tokio::spawn(async move {
+            // Used to avoid sending the same batch multiple times
+            let mut pending_batches = HashSet::new();
+            // let pair = cvar_pair_clone.clone();
```

Comment:
remove this commented line if it is not needed.

Reply:
Done!